pkg/store/cache/caching_bucket.go

```diff
@@ -132,35 +221,221 @@ func (cb *CachingBucket) ReaderWithExpectedErrs(expectedFunc objstore.IsOpFailur
+	}
+	return err
+}
+
+// Iter results should compress nicely, especially in subdirectories.
+func encodeIterResult(files []string) []byte {
+	data, err := json.Marshal(files)
+	if err != nil {
+		return nil
+	}
+
+	return snappy.Encode(nil, data)
+}
+
+func decodeIterResult(data []byte) ([]string, error) {
+	decoded, err := snappy.Decode(nil, data)
+	if err != nil {
+		return nil, err
+	}
+
+	var list []string
+	err = json.Unmarshal(decoded, &list)
+	return list, err
+}
+
+func (cb *CachingBucket) Exists(ctx context.Context, name string) (bool, error) {
+	cfgName, cache, cfg := cb.findCacheConfig(cb.operationConfigs[operationExists], name)
+	if cfg == nil {
+		return cb.Bucket.Exists(ctx, name)
+	}
+	existsCfg := cfg.(existsConfig)
+
+	cb.operationRequests.WithLabelValues(operationExists, cfgName).Inc()
+
+	key := cachingKeyExists(name)
+	hits := cache.Fetch(ctx, []string{key})
+
+	if ex := hits[key]; ex != nil {
+		switch string(ex) {
+		case existsTrue:
+			cb.operationHits.WithLabelValues(operationExists, cfgName).Inc()
+			return true, nil
+		case existsFalse:
+			cb.operationHits.WithLabelValues(operationExists, cfgName).Inc()
+			return false, nil
+		default:
+			level.Warn(cb.logger).Log("msg", "unexpected cached 'exists' value", "val", string(ex))
+		}
+	}
+
+	existsTime := time.Now()
+	ok, err := cb.Bucket.Exists(ctx, name)
+	if err == nil {
+		storeExistsCacheEntry(ctx, key, ok, existsTime, cache, existsCfg.existsTTL, existsCfg.doesntExistTTL)
+	}
+
+	return ok, err
+}
+
+func storeExistsCacheEntry(ctx context.Context, cachingKey string, exists bool, ts time.Time, cache cache.Cache, existsTTL, doesntExistTTL time.Duration) {
+	var (
+		data []byte
+		ttl  time.Duration
+	)
+	if exists {
+		ttl = existsTTL - time.Since(ts)
+		data = []byte(existsTrue)
+	} else {
+		ttl = doesntExistTTL - time.Since(ts)
+		data = []byte(existsFalse)
+	}
+
+	if ttl > 0 {
+		cache.Store(ctx, map[string][]byte{cachingKey: data}, ttl)
+	}
+}
+
+func (cb *CachingBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	cfgName, cache, cfg := cb.findCacheConfig(cb.operationConfigs[operationGet], name)
+	if cfg == nil {
+		return cb.Bucket.Get(ctx, name)
+	}
+	getCfg := cfg.(getConfig)
+
+	cb.operationRequests.WithLabelValues(operationGet, cfgName).Inc()
+
+	key := cachingKeyContent(name)
+	existsKey := cachingKeyExists(name)
+
+	hits := cache.Fetch(ctx, []string{key, existsKey})
+	if hits[key] != nil {
+		cb.operationHits.WithLabelValues(operationGet, cfgName).Inc()
+		return ioutil.NopCloser(bytes.NewReader(hits[key])), nil
+	}
+
+	// If we know that file doesn't exist, we can return that. Useful for deletion marks.
+	if ex := hits[existsKey]; ex != nil && string(ex) == existsFalse {
+		cb.operationHits.WithLabelValues(operationGet, cfgName).Inc()
+		return nil, errObjNotFound
+	}
+
+	getTime := time.Now()
+	reader, err := cb.Bucket.Get(ctx, name)
+	if err != nil {
+		if cb.Bucket.IsObjNotFoundErr(err) {
+			// Cache that object doesn't exist.
```

Comment:
small typo: Cache object that doesn't exist.

Reply:
I think it's correct the way it is. We want to cache information that object doesn't exist here.

Reply:
Ah got it.